src/hooks/useTektonResults.ts

```diff
@@ -0,0 +1,110 @@
+import React from 'react';
+import { K8sResourceCommon } from '@openshift/dynamic-plugin-sdk-utils';
+import { PipelineRunKind, TaskRunKind } from '../types';
+import {
+  getPipelineRuns,
+  TektonResultsOptions,
+  RecordsList,
+  getTaskRuns,
+  getTaskRunLog,
+} from '../utils/tekton-results';
+
+export type GetNextPage = () => void | undefined;
+
+const useTRRuns = <Kind extends K8sResourceCommon>(
+  getRuns: (
+    namespace: string,
+    options?: TektonResultsOptions,
+    nextPageToken?: string,
+    cacheKey?: string,
+  ) => Promise<[Kind[], RecordsList]>,
+  namespace: string,
+  options?: TektonResultsOptions,
+  cacheKey?: string,
+): [Kind[], boolean, unknown, GetNextPage] => {
+  const [nextPageToken, setNextPageToken] = React.useState<string>(null);
+
+  const [result, setResult] = React.useState<[Kind[], boolean, unknown, GetNextPage]>([
+    [],
+    false,
+    undefined,
+    undefined,
+  ]);
+
+  // reset token if namespace or options change
+  React.useEffect(() => {
+    setNextPageToken(null);
+  }, [namespace, options]);
+
+  React.useEffect(() => {
+    if (namespace) {
+      (async () => {
+        try {
+          const tkPipelineRuns = await getRuns(namespace, options, nextPageToken, cacheKey);
+          const token = tkPipelineRuns[1].nextPageToken;
+          setResult((cur) => [
+            nextPageToken ? [...cur[0], ...tkPipelineRuns[0]] : tkPipelineRuns[0],
+            true,
+            undefined,
+            token
+              ? (() => {
+                  // ensure we can only call this once
+                  let executed = false;
+                  return () => {
+                    if (!executed) {
+                      executed = true;
+                      // trigger the update
+                      setNextPageToken(token);
+                      return true;
+                    }
+                    return false;
+                  };
+                })()
+              : undefined,
+          ]);
+        } catch (e) {
+          if (nextPageToken) {
+            setResult((cur) => [cur[0], cur[1], e, undefined]);
+          } else {
+            setResult([[], false, e, undefined]);
+          }
+        }
+      })();
+    }
+  }, [namespace, options, nextPageToken, cacheKey, getRuns]);
```

Comment:
Need a destructor to set a `canceled` flag so as to not update state if effect is destroyed.